src/core/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,318 @@
+    if (
+      !inputValue ||
+      !isOpen ||
+      selectedOption?.label === inputValue // if a value is selected, show whole list to allow selecting a different value
+    ) {
+      setFilteredOptions(options);
+      return;
+    }
+    const _filteredOptions =
+      filterFunction?.(options, inputValue) ??
+      options.filter((option) =>
+        option.label.toLowerCase().includes(inputValue?.trim().toLowerCase()),
+      );
+    setFilteredOptions(_filteredOptions);
+    if (!_filteredOptions.find(({ value }) => value === selectedValue)) {
+      setSelectedValue(undefined);
+    }
+
+    setFocusedIndex((previouslyFocusedIndex) => {
+      if (_filteredOptions.includes(options[previouslyFocusedIndex])) {
+        return previouslyFocusedIndex;
+      } else if (
+        _filteredOptions.find(({ value }) => value === selectedValue)
+      ) {
+        return options.findIndex(({ value }) => value === selectedValue);
+      } else {
+        return -1; // reset focus if previously focused or selected value is not in filtered list
+      }
+    });
+  }, [inputValue, options, selectedValue, isOpen, filterFunction]);
+
+  const onKeyDown = React.useCallback(
+    (event: React.KeyboardEvent) => {
+      const focusedIndexInFilteredList = filteredOptions.findIndex(
+        (option) => option === options[focusedIndex],
+      );
+      switch (event.key) {
+        case 'ArrowDown':
+          if (isOpen) {
+            const nextIndex = Math.min(
+              focusedIndexInFilteredList + 1,
+              filteredOptions.length - 1,
+            );
+            setFocusedIndex(options.indexOf(filteredOptions[nextIndex]));
+          } else {
+            setIsOpen(true); // reopen menu if closed when typing
+          }
+          event.preventDefault();
+          event.stopPropagation();
+          break;
+        case 'ArrowUp':
+          if (isOpen) {
+            const previousIndex = Math.max(focusedIndexInFilteredList - 1, 0);
+            setFocusedIndex(options.indexOf(filteredOptions[previousIndex]));
+          }
+          event.preventDefault();
+          event.stopPropagation();
+          break;
+        case 'Enter':
+          isOpen && setSelectedValue(options[focusedIndex].value);
+          setIsOpen(!isOpen);
+          event.preventDefault();
+          event.stopPropagation();
+          break;
+        case 'Escape':
+          setIsOpen(false);
+          event.preventDefault();
+          event.stopPropagation();
+          break;
+        case 'Tab':
+          setIsOpen(false);
+          break;
+        default:
+          !isOpen && setIsOpen(true); // reopen menu if closed when typing
+          break;
+      }
+    },
+    [filteredOptions, focusedIndex, isOpen, options],
+  );
+
+  const menuItems = React.useCallback(
+    (close: () => void) => {
+      return filteredOptions.map((option) => {
+        const index = options.findIndex(({ value }) => option.value === value);
```

Comment:
I see everywhere you are using `options` instead of `filteredOptions` and then searching for the index inside `options`. Can you just use `filteredOptions` everywhere instead and `focusedIndex` just be current index?

Reply:
The big problem with that is that the id is tied to the index. And id of the focused option is used for setting activedescandant.